pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,462 @@
+	expectedVolumes []*api.PersistentVolume
+	// Initial content of controller claim cache.
+	initialClaims []*api.PersistentVolumeClaim
+	// Expected content of controller claim cache at the end of the test.
+	expectedClaims []*api.PersistentVolumeClaim
+	// Function to call as the test.
+	test testCall
+}
+
+type testCall func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error
+
+const testNamespace = "default"
+const versionConflictError = "VersionError"
+
+var noVolumes []*api.PersistentVolume
+var noClaims []*api.PersistentVolumeClaim
+
+// volumeReactor is a core.Reactor that simulates etcd and API server. It
+// stores:
+// - Latest version of claims volumes saved by the controller.
+// - Queue of all saves (to simulate "volume/claim updated" events). This queue
+//   contains all intermediate state of an object - e.g. a claim.VolumeName
+//   is updated first and claim.Phase second. This queue will then contain both
+//   updates as separate entries.
+// - Number of changes since the last call to volumeReactor.syncAll().
+// - Optionally, volume and claim event sources. When set, all changed
+//   volumes/claims are sent as Modify event to these sources. These sources can
+//   be linked back to the controller watcher as "volume/claim updated" events.
+type volumeReactor struct {
+	volumes              map[string]*api.PersistentVolume
+	claims               map[string]*api.PersistentVolumeClaim
+	changedObjects       []interface{}
+	changedSinceLastSync int
+	ctrl                 *PersistentVolumeController
+	volumeSource         *framework.FakeControllerSource
+	claimSource          *framework.FakeControllerSource
+	lock                 sync.Mutex
+}
+
+// React is a callback called by fake kubeClient from the controller.
+// In other words, every claim/volume change performed by the controller ends
+// here.
+// This callback checks versions of the updated objects and refuse those that
+// are too old (simulating real etcd).
+// All updated objects are stored locally to keep track of object versions and
+// to evaluate test results.
+// All updated objects are also inserted into changedObjects queue and
+// optionally sent back to the controller via its watchers.
+func (r *volumeReactor) React(action core.Action) (handled bool, ret runtime.Object, err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	glog.V(4).Infof("reactor got operation %q on %q", action.GetVerb(), action.GetResource())
+
+	switch {
+	case action.Matches("update", "persistentvolumes"):
+		obj := action.(core.UpdateAction).GetObject()
+		volume := obj.(*api.PersistentVolume)
+
+		// Check and bump object version
+		storedVolume, found := r.volumes[volume.Name]
+		if found {
+			storedVer, _ := strconv.Atoi(storedVolume.ResourceVersion)
+			requestedVer, _ := strconv.Atoi(volume.ResourceVersion)
+			if storedVer > requestedVer {
+				return true, obj, fmt.Errorf(versionConflictError)
```

Comment:
Proper error is VersionConflict defined here: https://github.com/jsafrane/kubernetes/blob/devel/refactor-binder/pkg/api/errors/errors.go#L156

However it's much harder to create than plain string and I am simply being lazy here. The error is opaque to the controller and it will bubble through the controller back to the test where I check the error string instead of `IsConflict(err)`. Does it really matter?